RFD-0000: Proposal to Include RFDs in Future System Changes (#1223)
cloudfoundry-incubator/kubecf: doc/rfd/0000/README.md

---
authors: Jaime Gomes <jaime.gomes@suse.com>
state: discussion
---

# Request for Discussion

The goal is to have an constructive discussion within the team about an idea and capture it in a
document known by RFD.

## Scope

## Metadata

1. authors: owners of the RFD must be listed with their name and email address 
(John Doe <john.doe@wonder.land>).
2. state: one of the states discussed below.

## State

1. draft
2. discussion
3. published
4. commited
5. abandoned

The ```draft``` state indicates that the work is not ready for discussion and that a placeholder

> was set. The RFD owner is puting his thoughs in place before promoting to ```discussion```.

Reviewer: ```suggestion
was set. The RFD owner is putting his thoughts in place before promoting to ```discussion```.
```

Optional gender-neutrality change (since I'm making suggestions on this line anyway):
```suggestion
was set. The RFD owner is putting their thoughts in place before promoting to ```discussion```.
```

Documents under active discussion within the team should be in ```discussion``` state. When reaching
this state the scope of the RFD must be clear and well defined.

Once the discussion converges to an agreement then the state should be set to ```published```.

Once well defined, the state should be in ```commited``` state. Any comment in the ```commited```
state should be raised as issues or if the comment brings fundamental changes to the original
content of the RFD then a new RFD should be created.

Finally, if an idea will never be implemented or it should be ignored then it can be moved to the
```abandoned``` state at any time.

## Workflow

Some of these steps may have some script support in the future to make everyones life easier.

### Find Next RFD Number

> Check for the next RFD number by going through the RFDs table.

Reviewer: This + next step can generate a template, maybe?

Author: I am trying to avoid having templates but probably we could have one just to fill the metadata.

### Create a Branch & Placeholder

Create a RFD branch using the name after the last (i.e. 0009) known RFD number:

``` shell
# git checkout -b rfd-0010
```

and create the placeholder:

``` shell
# mkdir -p docs/rfd/0010
# touch docs/rfd/0010/README.md
```

Fill in the RFD authors and set the state to ```draft```.

### Push Branch Remotely

Push the changes to your RFD branch.

``` shell
# git add docs/rfd/0010/README.md
# git commit -m "rfd-0010: adding placeholder for RFD..."
# git push origin 0010
```

In the future, after the branch pushed, the table in the README will be automatically updated to
reflect the new RFD or the change of state as well.

Now is time for you to fill the RFD with your thoughts to the point where you think it's good to
start sharing with others, to get their feedback and have a productive discussions.

### Open a Pull Request for Discussion

When ready to share with the team, then change the state from ```draft``` to ```discussion```
and push it remotely.

```shell
# git commit -m "rfd-0010: change RFD state to discussion"
# git push origin rfd-0010
```

Do not forget to open a pull request (maybe later we will have a bot for this) with the title

> _RDF:num_ .

Reviewer: Having a bot doesn't sound like a great idea, for thing in the `draft` stage.

Actually, why do we have a separate `draft` stage, since people can push things in branches (I personally prefer in forks) in `discussion`, and just don't actually open things for discussion by making a PR?

Author: @mook-as the RFD owner should open a PR when ready for ```discussion``` and the ```draft``` as a placeholder and to keep owner collected information to that point safe from loosing.

Author: on ```master``` will go only when the state is moved to ```published```. Until then, it will stay on the RFD branch.

With the pull request open anyone register to the repository can read the RFD and give feedback
through pull request comments.

It's up to the author(s) to incorporate any pull request comment into the RFD.

### Merge the Pull Request

After a while, and with team members feedback, the author(s) would merge the pull request into the

> master branch and change the state to ```published```.

Reviewer: Presumably we still need somebody to stamp it with a PR approval so GitHub would allow it? (Not the same person as the author(s), for GitHub workflow reasons… and good as a check that there actually was discussion, not that the check would normally be required.)

The state change needs to happen _before_ the merge, though, probably? (Because once it's merged we'd need a _second_ PR to make that change.)

Author: yes, the state must be changed before the merge (made some changes)

The timeline is up to the author(s), but as guideline it could go from a 3-5 days to 2 weeks.

> Before merging the pull request the state must also change to ```commited``` that indicates that the

Reviewer: I'm totally confused about the difference between `published` and `committed` now. What is the difference?  Is `published` a transient state between PR approval and merge? What actions must be taken to go from `published` to `committed`, in terms of things a person would actually have to do?

Ideally we would have a state transition diagram, with the transitions between states having captions describing the conditions for the transition (and possibly the actions taken during the transition).  But we probably don't need all that work?

Author: ```published``` means that the discussion ended with a clear understanding of what needs to be done to accomplish the initial idea. 

The ```committed``` indicates that the owners have finish the RFD and it's ready for implementation. I am leaving a blank space after this step since it's not the purpose of the RFD life-cycle to define.

Reviewer 2: Note that in my mind the phrases ```ended with a clear understanding of what needs to be done ...``` and ```owners have finish(ed) the RFD``` are the __same thing__. 

What is the difference between these two in your mind ?

Author: @andreas-kupries the idea is to have the ```published``` state like the end of the conversation but still requires some polish (grammar, indentation, ...) before moving to the ```committed``` state. I'll try to rephrase it....

Reviewer 2: I see. I have to admit I am not sure if that requires different states. We are already saying that any minor issues in the RFD will be handled via tickets, and polishing like that are minor, IMHO.

Wrt rephrasing, see also https://github.com/cloudfoundry-incubator/kubecf/pull/1223#discussion_r472393103 which is my take on the whole state section.

> RFD has involve from an idea to a clear description of a system or process and it can be

Reviewer: ```suggestion
RFD has evolved from an idea to a clear description of a system or process and it can be
```

Shouldn't things not be `published` _until_ we have a clear, shared understanding of what we're doing?  Isn't that the whole point of the discussion (as in `RFD`)?

---
authors: Jaime Gomes <jaime.gomes@suse.com>
state: discussion
---

# Request for Discussion

The goal is to have an constructive discussion within the team about an idea and capture it in a
document known by RFD.

## Scope

## Metadata

1. authors: owners of the RFD must be listed with their name and email address 
(John Doe <john.doe@wonder.land>).
2. state: one of the states discussed below.

## State

1. draft
2. discussion
3. published
4. committed
5. abandoned

The ```draft``` state indicates that the work is not ready for discussion and that a placeholder
was set. The RFD owner is putting his thoughts in place before promoting to ```discussion```.

> Documents under active discussion within the team should be in ```discussion``` state. When reaching

Reviewer 2: I am concerned by the use of `should be` here, and in the following state descriptions.
Here it leaves open the possibility of actively discussing a document which is still in `draft` state.
Similar for the other states.
And that means (I believe) what the various states are used for becomes a bit smeared in the mental landscape of the participants.

I will now try to come up with an alternate proposal for phrasing and describing the states where we use `must be` or `has to be` to proper delineate the differences between states.

Maybe this will even help with Mark's confusion. No promises though. As I may be confused as well, no ?

Reviewer 2: Here is my rewrite of this entire section:

```
The state ```draft``` indicates that the work is not ready for discussion yet.
It is a placeholder for the RFD owner to put his thoughts into.
The RFD owner has to state the scope of the RFD as clearly and concise as possible before promoting it to ```discussion```.

A work in the ```draft``` state may be exposed to people other than the owner.

The state ```discussion``` indicates that the work is ready for sharing and discussion.
When reaching this state the scope of the RFD must be clear and well defined.
At that point the RFD must be exposed to people other than the owner.
This should be at least the team itself, and must include relevant external stakeholders as well.

The state ```published``` indicates that the discussion has completed, and converged to a clear understanding of what needs to be done.
At that point the RFD is official and has to be made visible in the main development branch of the relevant repository.

From then on minor comments must be raised as issues of the repository.
Comments making fundamental changes to the RFD however have to restart the RFD process with a new RFD intended to supercede the RFD they comment on.

The last state, ```abandoned```, indicates that the RFD will never be implemented.
This state and can be reached from all previous states.
For a draft RFD the owner makes the descision to abandon it.
For an RFD under discussion the owner makes the descision, with guidance from the participants.
This should especially be considered when the discussion does not converge, indicating fundamental problems in the RFD, be it scope or understanding.
For a published RFD higher levels in and above the team make the decision.
```

Note, I left ```committed``` as I did not understand the difference to ```published```, even with the latest explanation from @fargozhu . Will comment there also.

Author: I did add some of your comments into the section but some I left 'cause I would prefer to avoid "mechanization" of the process.

Reviewer 2: I was less thinking `mechanization` and more `formal speaking`.
Looking ... Looks good. ... Actually like individual vs collective decision. Describes the difference without being strict on the collective.

Create a RFD branch using the name after the last (i.e. 0009) known RFD number:

``` shell
# git checkout -b rfd-0010
```

and create the placeholder:

``` shell
# mkdir -p docs/rfd/0010
# touch docs/rfd/0010/README.md
```

Fill in the RFD authors and set the state to ```draft```.

### Push Branch Remotely

Push the changes to your RFD branch.

``` shell
# git add docs/rfd/0010/README.md
# git commit -m "rfd-0010: adding placeholder for RFD..."
# git push origin 0010
```

In the future, after the branch pushed, the table in the README will be automatically updated to
reflect the new RFD or the change of state as well.

Now is time for you to fill the RFD with your thoughts to the point where you think it's good to
start sharing with others, to get their feedback and have a productive discussions.

### Open a Pull Request for Discussion

When ready to share with the team, then change the state from ```draft``` to ```discussion```
and push it remotely.

```shell
# git commit -m "rfd-0010: change RFD state to discussion"
# git push origin rfd-0010
```

Do not forget to open a pull request (maybe later we will have a bot for this) with the title
_RDF:num_ .

> With the pull request open anyone register to the repository can read the RFD and give feedback

Author: it should be _subscribed_ instead ;)

> With the pull request open anyone register to the repository can read the RFD and give feedback

Author: I was avoiding "state machine" to not overcomplicate the all idea and leave some space for personal decisions. But it would be something like this (after last changes to the text):

```
+---------+           +--------------+            +---------------+
|         |           |              |            |               |
|  draft  +---------->+  discussion  +----------->+   published   |
|         |   PR      |              |    Merge   |               |
+---+-----+           +-----+--------+            +---------------+
    |                       |
    |                       | Merge
    |                       |
    |                       v
    |              +--------+-------+
    |              |                |
    +------------->+   abandoned    |
                   |                |
      PR + Merge   +----------------+
```

> With the pull request open anyone register to the repository can read the RFD and give feedback

Reviewer 2: Looks good to me.

> With the pull request open anyone register to the repository can read the RFD and give feedback

Reviewer: Right, I was trying avoid having room for personal decisions, as much as reasonable.  That means I can do less thinking (around the process) at the time I create a RFD, since all of that thinking would be done ahead of time (i.e. right now).

Additionally, I find that having a diagrams helps _immensely_ towards figuring out what the actual process is.

through pull request comments.

It's up to the author(s) to incorporate any pull request comment into the RFD.

### Merge the Pull Request

> After a while, and with team members feedback, the author(s) would merge the pull request into the

Author: and that's fine!

> master branch, but before that it must the state to ```published```.

Reviewer: Workflow-wise, it seems easier to just classify anything not in `master` as not published? Why do we need the manual text-changing step, when the act of merging to `master` seems like a much better indication of state?

Author: There's some confusion from my side that requires some rephrasing. It should be merged when we switch to ```committed``` state and not to the ```published```.

Author: Better...I am going to propose the drop of the ```committed``` state and switch to ```published``` when merging.

The timeline is up to the author(s), but as guideline it could go from a 3-5 days to 2 weeks.

> Before merging the pull request the state must also change to ```committed``` that indicates that the

Reviewer: So I'd need to make two commits to the PR branch (`discussion` → `published`, `published` → `committed`) immediately before merging the PR? And all three commits (2 text changes + 1 merge commit) are triggered by the same thing (PR approval / consensus)?

Author: ```committed``` state is gone.

(John Doe <john.doe@wonder.land>).
2. state: one of the states discussed below.

## State

1. draft
2. discussion
3. published
4. abandoned

The ```draft``` state indicates that the work is not ready for discussion yet. It is a placeholder
for the RFD author(s) to put their thoughts into.
The RFD author(s) has to state the scope of the RFD as clearly and concise as possible before
promoting it to ```discussion```.

Documents under active discussion within the team must be in ```discussion``` state. When reaching
this state the scope of the RFD must be clear and well defined.

Once the discussion converges to clear understanding of what needs to be done, then the state must
be set to ```published```.

Any comment in the ```published``` state must be raised as issues or if the comment brings
fundamental changes to the original content of the RFD then a new RFD must be created.

The ```abandoned``` state indicates that the RFD will not be implemented and it can be switched from
any previous state. Depending of the current RFD state, it may be the author(s) making an individual
decision (```draft```) or a collective decision (```discussion``` or ```published```).

## Workflow

> Some of these steps may have some script support in the future to make everyones life easier.

Reviewer: ```suggestion
Some of these steps may have some script support in the future to make everyone's life easier.
```

### Find Next RFD Number

Check for the next RFD number by going through the RFDs table.

### Create a Branch & Placeholder

Create a RFD branch using the name after the last (i.e. 0009) known RFD number:

``` shell
# git checkout -b rfd-0010
```

and create the placeholder:

``` shell
# mkdir -p docs/rfd/0010
# touch docs/rfd/0010/README.md
```

Fill in the RFD authors and set the state to ```draft```.

### Push Branch Remotely

Push the changes to your RFD branch.

``` shell
# git add docs/rfd/0010/README.md
# git commit -m "rfd-0010: adding placeholder for RFD..."
# git push origin 0010
```

In the future, after the branch pushed, the table in the README will be automatically updated to
reflect the new RFD or the change of state as well.

Now is time for you to fill the RFD with your thoughts to the point where you think it's good to

> start sharing with others, to get their feedback and have a productive discussions.

Reviewer: ```suggestion
start sharing with others, to get their feedback and have a productive discussion.
```

### Open a Pull Request for Discussion

When ready to share with the team, then change the state from ```draft``` to ```discussion```
and push it remotely.

```shell
# git commit -m "rfd-0010: change RFD state to discussion"
# git push origin rfd-0010
```

Do not forget to open a pull request (maybe later we will have a bot for this) with the title
_RDF:num_ .

With the pull request open anyone subscribed to the repository will be notified about the new RFD

> and it can givethe feedback through pull request comments.

Reviewer: ```suggestion
and it can gather feedback through pull request comments.
```

It's up to the author(s) to incorporate any pull request comment into the RFD.

### Merge the Pull Request

After a while, and with team members feedback, the author(s) would merge the pull request into the
master branch, but before that it must the state to ```published```. At this moment, the RFD must be
ready for implementation.

If an RFD state is switched to ```abandoned``` the pull request must be merged into the master

> branch. If the RFD is in a ```draft``` state then a pull request must be open and merged for future reference.

Reviewer: ```suggestion
branch. If the RFD is in a ```draft``` state then a pull request must be opened and merged for future reference.
```

Practically speaking, this means we still need to involve somebody else to abandon a RFD?  I wonder if it would make more sense for RFDs that never reach the `discussion` stage to be abandoned without merging.
